src/services/download/sentry.rs

```diff
@@ -219,37 +219,60 @@ impl SentryDownloader {
         file_source: SentryObjectFileSource,
         destination: PathBuf,
     ) -> Result<DownloadStatus, DownloadError> {
-        let download_url = file_source.url();
-        log::debug!("Fetching debug file from {}", download_url);
-        let result = future_utils::retry(|| {
-            self.client
-                .get(download_url.clone())
-                .header("User-Agent", USER_AGENT)
-                .header(
-                    "Authorization",
-                    format!("Bearer {}", &file_source.source.token),
-                )
-                .send()
-        });
-
-        match result.await {
+        match future_utils::retry(|| {
+            self.download_source_once(file_source.clone(), destination.clone())
+        })
+        .await
+        {
+            Ok(status) => {
+                log::debug!(
+                    "Fetched debug file from {}: {:?}",
+                    file_source.url(),
+                    status
+                );
+                Ok(status)
+            }
+            Err(err) => {
+                log::debug!(
+                    "Failed to fetch debug file from {}: {}",
+                    file_source.url(),
+                    err
+                );
+                Err(err)
+            }
+        }
+    }
+
+    async fn download_source_once(
+        &self,
+        source: SentryObjectFileSource,
+        destination: PathBuf,
+    ) -> Result<DownloadStatus, DownloadError> {
+        match self
+            .client
+            .get(source.url())
+            .header("User-Agent", USER_AGENT)
+            .bearer_auth(&source.source.token)
+            .send()
+            .await
+        {
             Ok(response) => {
                 if response.status().is_success() {
-                    log::trace!("Success hitting {}", download_url);
+                    log::trace!("Success hitting {}", source.url());
```

Comment:
These calls to `url()` allocate each time. Can we keep this deduplicated as before?

Reply:
They don't duplicate allocations though.  Each branch calls it only once, and `self.client.get()` consumes anyway so you need to construct it twice in all cases.  (I originally had it in a variable as you suggested but it seemed noisier for no gain)

Reply:
That's right. I should've put this comment on the retry closure.

`IntoUrl` from reqwest is implemented for `&str` (since it needs to construct an owned `Url` object anyway), so we could've just passed a reference by ensuring that `utils::retry` accepts an `FnMut`. Or, potentially even better, parse the URL once and then pass a clone (or reference?) to `get`.

Reply:
to fill in an offline conversation: `.url()` already returns `Url` type so the gains are very marginal because the retry can not be triggered by the wrong url type